common/archiver/archivalMetadata.go

```diff
@@ -148,34 +123,53 @@ func NewArchivalConfig(
 	}
 
 	return &archivalConfig{
-		clusterStatus:       clusterStatus,
-		enableRead:          enableRead,
-		domainDefaultStatus: domainDefaultStatus,
-		domainDefaultURI:    domainDefaultURI,
+		staticClusterStatus:  staticClusterStatus,
+		dynamicClusterStatus: dc.GetStringProperty(dynamicconfig.HistoryArchivalStatus, staticClusterStatusStr),
+		enableRead:           dc.GetBoolProperty(dynamicconfig.EnableReadFromHistoryArchival, enableRead),
+		domainDefaultStatus:  domainDefaultStatus,
+		domainDefaultURI:     domainDefaultURI,
 	}
 }
 
 // NewDisabledArchvialConfig returns a disabled ArchivalConfig
 func NewDisabledArchvialConfig() ArchivalConfig {
 	return &archivalConfig{
-		clusterStatus:       ArchivalDisabled,
-		enableRead:          false,
-		domainDefaultStatus: shared.ArchivalStatusDisabled,
-		domainDefaultURI:    "",
+		staticClusterStatus:  ArchivalDisabled,
+		dynamicClusterStatus: nil,
+		enableRead:           nil,
+		domainDefaultStatus:  shared.ArchivalStatusDisabled,
+		domainDefaultURI:     "",
 	}
 }
 
 // ClusterConfiguredForArchival returns true if cluster is configured to handle archival, false otherwise
 func (a *archivalConfig) ClusterConfiguredForArchival() bool {
-	return a.clusterStatus == ArchivalEnabled
+	return a.GetClusterStatus() == ArchivalEnabled
 }
 
 func (a *archivalConfig) GetClusterStatus() ArchivalStatus {
-	return a.clusterStatus
+	// Only check dynamic config when archival is enabled in static config.
+	// If archival is disabled in static config, there will be no provider section in the static config
+	// and the archiver provider can not create any archiver. Therefore, in that case,
+	// even dynamic config says archival is enabled, we should ignore that.
+	// Only when archival is enabled in static config, should we check if there's any difference between static config and dynamic config.
+	if a.staticClusterStatus != ArchivalEnabled {
+		return a.staticClusterStatus
+	}
+
+	dynamicStatusStr := a.dynamicClusterStatus()
+	dynamicStatus, err := getClusterArchivalStatus(dynamicStatusStr)
+	if err != nil {
+		dynamicStatus = a.staticClusterStatus
```

Comment:
I think that if we ever fail to read dynamic config we should always default to disabled state even if archival config is enabled. Thoughts?

Reply:
I can argue both ways and I don't have a strong opinion on this. so changed to disabled when dynamic config is invalid.